Scopes for bigger organizations

leio10 opened the pull request:
#### :tophat: What? Why?
This PR improves scopes system. It migrates existing data to the new schema, allows to administrate hierarchically and adapt built-in engines to use it.

#### :pushpin: Related Issues
- Implements #1480

#### :clipboard: Subtasks
- [x] Schema and model changes 
- [x] Changes in the administration tool 
  - [x] Select2 translations
  - [x] Non-english translations
- [x] Changes in the front
  - [x] Clicable scope tags on resources allow to search by non-listed scopes
- [x] Tests adapted
- [x] Translations created

### :camera: Screenshots (optional)
![imagen](https://user-images.githubusercontent.com/453545/27227563-3d2d5f38-52a4-11e7-95af-46a286ea467a.png)

#### :ghost: GIF
![](https://media.giphy.com/media/SzkfCPdLnESS4/giphy.gif)

josepjaume commented on decidim-admin/app/views/decidim/admin/scopes/search_results.json.erb:
```diff
@@ -0,0 +1,9 @@
+<%
```
What about doing this on the controller layer (maybe via a presenter or serializer) instead? JSON is not markup language and doing this here looks a bit out of place (or maybe I'm missing something?)

leio10 replied:
My original solution used the `to_json` method, but I get stuck with the translatable attribute, so I ended up with this ugly solution. I think I will use ProposalSerializer or CommentSerializer as an example.

josepjaume commented on decidim-core/config/locales/en.yml:
```diff
@@ -276,6 +276,7 @@ en:
   errors:
     messages:
       content_type_whitelist_error: The file type is not valid
+      cycle_detected: a scope descendant can't be its parent
```
shouldn't this be the other way around - `a scope's parent can't be one of its descendants`? Why are not hiding them though?

leio10 replied:
Yes, I'm afraid that the problem is the English, not the idea. :see_no_evil: So, I will change the text. 
Regarding the second question, this message is used on model validation, but I do not believe that scenario could happen using the current admin UI.

josepjaume commented on decidim-core/db/migrate/20170605162500_add_hierarchy_to_scopes.rb:
```diff
@@ -0,0 +1,60 @@
+    end
+
+    # post migration data fixes
+    Decidim::Scope.all.find_each do |s|
```
Maybe you can store all the data in an in-memory variable instead - doing it this way looks a bit hacky, doesn't it?

Wrap the whole thing inside a transaction and it's good to go!

leio10 replied:
Yes, it looks very hacky to me too. I've been searching for a more elegant solution, but I didn't find anything. Taking into account that current scopes are flat, I will assume that there are very few on current systems. So, if no new problems appear, I will load all of them on memory and do the migration. :+1:

oriolgual replied:
Since there are structure changes Rails already wraps all this in a transaction.

mrcasals commented on decidim-admin/app/assets/stylesheets/decidim/admin/modules/_secondary-nav.scss:
```diff
@@ -51,7 +51,8 @@ $process-title-height: 3rem;
 .secondary-nav__title,
-.process-title-content{
+.process-title-content,
+.scope-title-content{
```
@josepjaume you mean `decidim-design`?

andreslucena replied:
For the record, now they live on:
https://github.com/decidim/design
https://github.com/decidim/design-admin

leio10 replied:
Ok, there are several changes on scopes section that should be reflected on that repo, and also minor changes on participatory processes and other engines. So, the first question is: the synchronization is done by hand? If so, I will fork both repos and submit a PR with the proposed changes and comment there the problems that I found.

leio10 replied:
I've created this PR with the admin changes: decidim/design-admin#22

leio10 commented:
Thank you very much for the feedback, @josepjaume! It is not nitpicky, on the contrary, I am very interested in learning and understanding how are you working. There go my answers...

leio10 commented on decidim-admin/app/forms/decidim/admin/scope_form.rb:
```diff
@@ -4,21 +4,29 @@ module Decidim
+      attribute :code, String
+      attribute :parent_id, Integer
+      attribute :scope_type_id, Integer
+      attribute :deprecated, Boolean
```
Yes, old processes should keep it original scope. Also, users related to that scope may be asked to choose another scope.
Thinking about it, I think my solution is incomplete. It marks deprecated territories on participatory process edition form, but it should also hide those territories in new processes, while keeping it on old ones. 
Maybe the best solution would be to change deprecated column to a deprecated_at datetime. That way, scope elegibility and searchability would depend on participatory process creation date. What do you think?

josepjaume replied:
This is probably a good idea, but I might be more comfortable discussing it in a separate feature. What do you think of shipping a MVP of this first (hierarchical scopes) and then refine it (deprecation, searchability, etc). What do you think? There's implications of this at the UI & UX level that I would feel more comfortable exploring better.

leio10 commented on decidim-admin/app/queries/decidim/admin/freetext_scopes.rb:
```diff
@@ -0,0 +1,22 @@
+module Decidim
+  module Admin
+    # This query searches scopes by name.
+    class FreetextScopes < Rectify::Query
```
Yes, I've been looking at that gem too, and I also reach the same conclusion too :smile:

oriolgual commented on decidim-admin/app/commands/decidim/admin/create_scope.rb:
```diff
@@ -7,8 +7,9 @@ class CreateScope < Rectify::Command
       #
       # form - A form object with the params.
-      def initialize(form)
+      def initialize(form, parent_scope = nil)
```
Add docs for parent scope

oriolgual commented on decidim-admin/app/commands/decidim/admin/create_scope.rb:
```diff
@@ -31,7 +32,11 @@ def call
-          organization: form.organization
+          organization: form.organization,
+          code: form.code,
+          scope_type_id: form.scope_type_id,
```
Can we use objects instead of IDs?

josepjaume replied:
Also: I think Scope Types should be optional. Eases the transition from our current data model, and makes sense when having arbitrary trees of scopes in scenarios where they don't have a clear categorization.

oriolgual commented on decidim-admin/app/controllers/decidim/admin/scopes_controller.rb:
```diff
@@ -74,9 +88,33 @@ def scope
         @scope ||= collection.find(params[:id])
       end
 
+      def parent_scope
```
Why do we need this and how it differs from `scope`?

leio10 replied:
I've changed the core to allow browsing the scopes tree with the same controller and views. So, I need to know if I'm showing a top level scope or a child. In the second case, this method gives me, the parent scope.

oriolgual commented on decidim-admin/app/helpers/decidim/admin/scopes_helper.rb:
```diff
@@ -0,0 +1,34 @@
+  module Admin
+    # This module includes helpers to show scopes in admin
+    module ScopesHelper
+      def scope_breadcrumbs(current_scope)
```
Add docs

oriolgual commented on decidim-admin/app/helpers/decidim/admin/scopes_helper.rb:
```diff
@@ -0,0 +1,34 @@
+        end
+      end
+
+      def process_scopes_for_select(participatory_process)
```
Add docs

oriolgual commented on decidim-admin/app/helpers/decidim/admin/scopes_helper.rb:
```diff
@@ -0,0 +1,34 @@
+          end
+      end
+
+      def organization_scope_types(organization = current_organization)
```
Add docs

oriolgual commented on decidim-admin/app/queries/decidim/admin/freetext_scopes.rb:
```diff
@@ -0,0 +1,22 @@
+  module Admin
+    # This query searches scopes by name.
+    class FreetextScopes < Rectify::Query
+      def self.for(organization, lang, name)
```
Add docs

oriolgual commented on decidim-admin/app/queries/decidim/admin/freetext_scopes.rb:
```diff
@@ -0,0 +1,22 @@
+        new(organization, lang, name).query
+      end
+
+      def initialize(organization, lang, name)
```
Add docs

oriolgual commented on decidim-admin/app/views/decidim/admin/scope_types/index.html.erb:
```diff
@@ -0,0 +1,40 @@
+          </tr>
+        </thead>
+        <tbody>
+          <% @scope_types.each do |scope_type| %>
```
Do we need this instance var, can't we use `collection`?

leio10 replied:
I don't understand this comment, all the admin controllers and index views do it the same way.

deivid-rodriguez replied:
I think @oriolgual means the `collection` method defined in the controller. I think this is used in the participatory process controller as some trick to render heterogeneous collections (process + process_groups) transparently. But in this case I think we don't need that. I would rename the `collection` method to `scope_types`, expose it as a helper and use that.

oriolgual commented on decidim-admin/app/views/decidim/admin/scopes/index.html.erb:
```diff
@@ -1,27 +1,28 @@
               </td>
               <td>
-                <%= l scope.created_at, format: :short %>
+                <%= translated_attribute(scope.scope_type.name) %>
```
What about *old* scopes that don't have a type?

oriolgual commented on decidim-core/app/models/decidim/organization.rb:
```diff
@@ -23,6 +24,10 @@ class Organization < ApplicationRecord
     mount_uploader :logo, Decidim::OrganizationLogoUploader
     mount_uploader :favicon, Decidim::OrganizationFaviconUploader
 
+    def top_scopes
```
Add docs

oriolgual commented on decidim-core/app/models/decidim/organization.rb:
```diff
@@ -23,6 +24,10 @@ class Organization < ApplicationRecord
     mount_uploader :favicon, Decidim::OrganizationFaviconUploader
 
+    def top_scopes
+      scopes.top_level
```
Should this be memoized? (@top_scopes ||= scopes.top_level)

oriolgual commented on decidim-core/app/models/decidim/participatory_process.rb:
```diff
@@ -74,5 +74,9 @@ def published?
       attributes["hashtag"].to_s.delete("#")
     end
+
+    def subscopes
```
Docs

oriolgual commented on decidim-core/app/models/decidim/scope.rb:
```diff
@@ -10,6 +10,70 @@ class Scope < ApplicationRecord
+                inverse_of: :children,
+                optional: true
+
+    has_many    :children,
```
I'd prefer having a single space between `has_many` and `children`, the same for above belongs to.

deivid-rodriguez replied:
I think this can be enforced by RuboCop, I can add it.

oriolgual commented on decidim-core/app/models/decidim/scope.rb:
```diff
@@ -10,6 +10,70 @@ class Scope < ApplicationRecord
+    validates :code, uniqueness: { scope: :organization }
+    validate :forbid_cycles
+    after_create_commit :create_part_of
+    before_validation :update_part_of, on: :update
```
Can you move this before the validations?

oriolgual commented on decidim-core/app/models/decidim/scope.rb:
```diff
@@ -10,6 +10,70 @@ class Scope < ApplicationRecord
+    validates :name, :code, :organization, presence: true
+    validates :code, uniqueness: { scope: :organization }
+    validate :forbid_cycles
+    after_create_commit :create_part_of
```
Can you add a new line between validates and this?

oriolgual commented on decidim-core/app/models/decidim/scope.rb:
```diff
@@ -10,6 +10,70 @@ class Scope < ApplicationRecord
+    end
+
+    def part_of_scopes
+      Decidim::Scope.where(id: part_of)
```
Redundant class?

leio10 replied:
It is not a scope, it is an instance method that retrieves all part_of scope objects.

oriolgual commented on decidim-core/app/models/decidim/scope.rb:
```diff
@@ -10,6 +10,70 @@ class Scope < ApplicationRecord
+      end
+    end
+
+    def part_of_scopes
```
Add docs

oriolgual commented on decidim-core/app/models/decidim/scope.rb:
```diff
@@ -10,6 +10,70 @@ class Scope < ApplicationRecord
+
+    def forbid_cycles
+      if parent && parent.part_of.include?(id)
+        errors.add(:parent_id, :cycle_detected)
```
One liner!

oriolgual commented on decidim-core/app/models/decidim/scope_type.rb:
```diff
@@ -0,0 +1,17 @@
+# frozen_string_literal: true
+
+module Decidim
+  # Scopes are used in some entities through Decidim to help users know which is
```
Updated docs (Scope Types?)

oriolgual commented on decidim-core/app/services/decidim/resource_search.rb:
```diff
@@ -39,12 +39,22 @@ def search_category_id
+
+      conditions = []
+      if clean_scope_ids.delete("global")
+        conditions << "decidim_scope_id IS NULL"
```
Convert to one liner

oriolgual commented on decidim-core/app/services/decidim/resource_search.rb:
```diff
@@ -39,12 +39,22 @@ def search_category_id
+      if clean_scope_ids.delete("global")
+        conditions << "decidim_scope_id IS NULL"
+      end
+      if clean_scope_ids.any?
```
Add new line

oriolgual commented on decidim-core/app/services/decidim/resource_search.rb:
```diff
@@ -39,12 +39,22 @@ def search_category_id
+      end
+      if clean_scope_ids.any?
+        self.query = query.joins("LEFT JOIN decidim_scopes ON decidim_scope_id=decidim_scopes.id")
+        conditions.concat ["? = ANY(decidim_scopes.part_of)"] * clean_scope_ids.count
```
I don't understand this

leio10 replied:
This method receives a list of scopes ids to find, maybe including "global" keyword for global resources.
On one hand, you could have to retrieve all resources whose scope in part of any of the scopes on the list.
On the other hand, you could have to retrieve resources with no scope, that is global.
If the user ask for both conditions, sentences are joined by an OR, so I can't use an inner join because it will exclude resources with no scope.

I don't like this solution very much; the right one would require that every process has a scope. That would means that "Global" scope should be the parent scope for every top level scopes, because [nothing is something](https://www.youtube.com/watch?v=OMPfEXIlTVE). But I'm not sure if it raises too much the complexity.

deivid-rodriguez replied:
Not sure about this specific feature since I haven't looked at it closely. But many times the right solution ends up being the simpler one too. It would certainly simplify the bit of code... :)

oriolgual commented on decidim-core/app/models/decidim/scope_type.rb:
```diff
@@ -0,0 +1,17 @@
+  # Scopes are used in some entities through Decidim to help users know which is
+  # the scope of a participatory process.
+  # (i.e. does it affect the whole city or just a district?)
+  class ScopeType < ApplicationRecord
```
Missing specs

oriolgual commented on decidim-core/spec/models/decidim/scope_spec.rb:
```diff
@@ -4,31 +4,63 @@
+    end
 
-        expect(other_scope).to be_valid
+    let(:subscope) { create(:subscope, parent: scope) }
```
Move lets after context declaration

oriolgual commented on decidim-core/spec/models/decidim/scope_spec.rb:
```diff
@@ -4,31 +4,63 @@
-      it "is valid" do
-        expect(scope).to be_valid
+    it { is_expected.to be_valid }
+
```
Add `part_of` specs

oriolgual commented on decidim-core/spec/models/decidim/scope_spec.rb:
```diff
@@ -4,31 +4,63 @@
 module Decidim
   describe Scope do
-    let(:scope) { build(:scope) }
+    let(:scope) { create(:scope) }
```
Use build when possible (if it weren't valid it would crash).

oriolgual commented on decidim-core/lib/decidim/core/test/factories.rb:
```diff
@@ -268,11 +276,29 @@
   end
 
+  factory :scope_type, class: Decidim::ScopeType do
+    name Decidim::Faker::Localized.word
```
Wrap in { }

leio10 commented:
Ok, I've made serveral things:
- created a PR for the admin design change (decidim/design-admin#22)
- removed scopes deprecation and metadata from this PR
- added missing docs and tests
- discovered and fixed some bugs

So, if all this is OK, there are only two pending issues:
- Translations: 
  - Select2 library has it own i18n system, with additional JS files for each language. Have you run into a situation like this before? If not, I will add reference for all languages .js after loading select2 library, to avoid increasing complexity.
  - I've only created English translations, it is ok that way? I haven't worked with Crowdin before, so I don't which is the work flow.
- Resources filters: 
  - Filters by scope are working as before, using only the next level of subscopes of the process, and it works OK. When showing resources results, it appears a tag with the resource scope, that could be two levels lower than the scope of the process. If you click on it, it also works, but it is not visible on filters list. I will create a PR on decidim/design repo to address this.

Thank you very much for the help, @josepjaume and @oriolgual! I think that I already have quite clear your way of working and I hope the next PR will require less effort. Don't worry about feelings, truth does not hurt. :relaxed:

deivid-rodriguez commented on decidim-core/app/services/decidim/resource_search.rb:
```diff
@@ -39,12 +39,21 @@ def search_category_id
+      clean_scope_ids = [scope_id].flatten
+
+      conditions = []
+      conditions << "decidim_scope_id IS NULL" if clean_scope_ids.delete("global")
```
I think the right check here would be to check the other table: `decidim_scopes: { id: nil }`. Otherwise, if there are any orphan records, they'll be included in the result. If there was a foreign key between from `decidim_scope_id` pointing to the `scopes` table it would be equivalent I think. But the standard way to do it is to check the joined table anyways.

leio10 replied:
Mmm, I'm not doing the join if the only condition is "global", that is why I use the feature column instead of the scope's column.

deivid-rodriguez replied:
Ah, I see. Makes sense although it makes me wonder if we should go with the "right" solution you proposed to avoid those conditionals here.

And we need to add that foreign key if it's not there already to ensure data integrity.

deivid-rodriguez commented on decidim-core/app/services/decidim/resource_search.rb:
```diff
@@ -39,12 +39,21 @@ def search_category_id
+      conditions << "decidim_scope_id IS NULL" if clean_scope_ids.delete("global")
+
+      if clean_scope_ids.any?
+        self.query = query.joins("LEFT JOIN decidim_scopes ON decidim_scope_id=decidim_scopes.id")
```
Rails now supports `left_joins` out of the box! :wink:

leio10 replied:
That is good news, thanks!

deivid-rodriguez commented on .gitignore:
```diff
@@ -14,3 +14,4 @@ karma-*
 yarn-error.log
 coverage
 webpack.report.html
+*.sublime-*
```
I think this kind of gitignore belongs in each developer's global gitignore. That way you add it to your global gitignore once in your life and you don't have to add it to the gitignore of every project you work with.

andreslucena commented:
>     Select2 library has it own i18n system, with additional JS files for each language. Have you run into a situation like this before? If not, I will add reference for all languages .js after loading select2 library, to avoid increasing complexity.

We're having the same situation with Datepicker JS library, and as of now we're putting the JS files manually: 
https://github.com/decidim/decidim/tree/master/decidim-core/vendor/assets/javascripts/datepicker-locales

There are two possiblities: 
a) hacking the .js files and making it compatible with crowdin (for instance, doing a .js.erb with a for loop with I18n.available_languages).
b) adding it to the documentation for a new language, saying: 

## Adding new language

If you want to add a new language to Decidim, you should: 
1- Ask us on a GitHub issue so we can add it on Crowdin
2- Create a user on Crowdin, wait until you have permission as a collaborator
3- When it's added on Crowdin start translating it 
4- Make a Pull Request on GitHub adding the files for this languages on the javascript libraries that we use: 
https://github.com/decidim/decidim/tree/master/decidim-core/vendor/assets/javascripts/datepicker-locales 
5- Review it on the demo site and correct the translations. 

>     I've only created English translations, it is ok that way? I haven't worked with Crowdin before, so I don't which is the work flow.

Yes, I think that's ok. If i'm not wrong, the flow is adding the strings manually on english, merging the PR and then translating them on crowdin.

josepjaume commented:
@leio10 super nice! I think we'll end up merging `decidim-design-admin` and `decidim-design` in the `decidim` repo so they use the same stylesheets but still live as a separate static site that works as a styleguide. This "send a PR to the design project" is becoming a burden, but it helped a lot in the past when the design was waaaay before us.

josepjaume commented:
Nice! @beagleknight can I ask for a review? 

Also @leio10 looks like we need to rebase.

josepjaume commented:
Hi @leio10, we're taking a good look at this. I've merged the latest changes on *master* and now we're checking it against a working app. If everything is OK, it's gonna get merged!

beagleknight commented on decidim-admin/app/models/decidim/admin/abilities/admin_ability.rb:
```diff
@@ -29,6 +29,7 @@ def define_abilities
           can :manage, Moderation
           can :manage, Attachment
           can :manage, Scope
+          can :manage, ScopeType
```
I think we should add the same ability to a `ParticipatoryProcessAdmin`, right?

beagleknight commented:
I have created a review app to help the review of this PR: https://decidim-localret-pr-8.herokuapp.com

beagleknight commented:
@decidim/developers

beagleknight commented:
Hi @leio10 !

I did some manually testing and I detected a few things:

- **production:** I'm getting a JS error when I try to filter by scope. You can reproduce this navigating to https://decidim-localret-pr-8.herokuapp.com/processes/3/f/9/ and trying to filter the collection by scope. The error is `Uncaught ReferenceError: Rails is not defined`. It looks like you are expecting a global variable but it is not present in the production build.

- **development:** When I navigate to http://localhost:3000/processes/1/f/7/proposals and try to filter the collection by the global scope it works fine. But if I select other scope I get an error: `PG::DuplicateAlias: ERROR: table name "decidim_scopes" specified more than once `. I tried this and it seems to work properly:

```git
diff --git a/decidim-core/app/services/decidim/resource_search.rb b/decidim-core/app/services/decidim/resource_search.rb
index 844b766..064f32a 100644
--- a/decidim-core/app/services/decidim/resource_search.rb
+++ b/decidim-core/app/services/decidim/resource_search.rb
@@ -49,7 +49,7 @@ module Decidim
       conditions << "decidim_scope_id IS NULL" if clean_scope_ids.delete("global")

       if clean_scope_ids.any?
-        self.query = query.left_outer_joins(:scope)
+        self.query = query.joins(:scope)
         conditions.concat ["? = ANY(decidim_scopes.part_of)"] * clean_scope_ids.count
       end
```

It also looks you are missing a feature spec here 😄 .

beagleknight commented:
@leio10 It looks like you can access the same `Rails` variable using `$.rails` but I am not really sure 😄

leio10 commented:
Hi @beagleknight, thanks for the feedback!
* Regarding the global variable, I didn't find good documentation for that. I've tried `$.rails`, but it didn't work on the development environment :disappointed:. When I've searched for `rails-ujs` it takes me to https://github.com/rails/jquery-ujs package. At this point, I think that the best solution is to choose the other solution that I proposed at #1607: replace `rails-ujs` library with the `jquery_ujs` in the development app generator. What do you think?
* Regarding the JOIN, it is necessary to use a LEFT JOIN because the scope of the participatory process could be NULL (when the process is related to the global scope). This code is working OK at my dev environment, so this solution doesn't give much confidence. Please, check decidim/decidim/pull/1500#discussion_r123462336: I'd create a scope for global scope instead of NULL, but this has many implications (and a lot of code to change :cry:). Maybe @josepjaume has anything to say about this. 
* Regarding the missing spec, I'm going to add it.

leio10 commented:
Hi again @beagleknight, I've been reviewing these problems, and I think there are all fixed:
* As I said, I've replaced `rails-ujs` with `jquery_ujs` and everything works OK, without needing global variables.
* Regarding the LEFT JOIN, I think the problem was that ProposalsController was calling include(:scope) on the search results relation. I changed the `left_outer_join` with `includes` and `references` and now works well.
* Regarding the missing spec, I think is very difficult to add tests for an abstract class. It is well tested through the child classes. I've added some extra tests to the ProposalsSearch class, to cover cases when there are subscopes and mixing global scope with other scopes.

beagleknight commented:
Hi there @leio10 !

> As I said, I've replaced rails-ujs with jquery_ujs and everything works OK, without needing global variables.

That's awesome! I really don't like global variables leaking everywhere so I am glad you found a better solution 😄 

> Regarding the LEFT JOIN, I think the problem was that ProposalsController was calling includes(:scope) on the search results relation. I changed the left_outer_join with includes and references and now works well.

Nice catch!

> Regarding the missing spec, I think is very difficult to add tests for an abstract class (ResourceSearch). It is well tested through the child classes. I've added some extra tests to the ProposalsSearch class, to cover cases when there are subscopes and mixing global scope with other scopes.

I think a good test could be adding a new one here: https://github.com/decidim/decidim/blob/master/decidim-proposals/spec/features/proposals_spec.rb and filtering proposals by scope. It wasn't there before so it's not completely related to your pull request. If you have time to add the spec, it would be great 😄 .

josepjaume commented:
Yeah, the thing that bothers me is that the test suite passed but the front-end side of things was broken. Let's add that spec and it's a GO.

josepjaume commented:
Thanks everyone for the hard work! ❤️

leio10 commented:
@beagleknight @josepjaume done! Thank you for the patience with my first PR :slightly_smiling_face:

beagleknight commented:
Hi @leio10 !

I am trying to recreate the review app for this branch but I think I found a problem in `20170719122332_add_hierarchy_to_scopes.decidim`:

```
undefined method `organization' for #<AddHierarchyToScopes:0x007f9b0c94b340>
/Users/beagleknight/.gem/ruby/2.4.1/gems/activerecord-5.1.2/lib/active_record/migration.rb:848:in `block in method_missing'
/Users/beagleknight/.gem/ruby/2.4.1/gems/activerecord-5.1.2/lib/active_record/migration.rb:818:in `block in say_with_time'
/Users/beagleknight/.gem/ruby/2.4.1/gems/activerecord-5.1.2/lib/active_record/migration.rb:818:in `say_with_time'
/Users/beagleknight/.gem/ruby/2.4.1/gems/activerecord-5.1.2/lib/active_record/migration.rb:838:in `method_missing'
/Users/beagleknight/.gem/ruby/2.4.1/gems/activerecord-5.1.2/lib/active_record/migration.rb:600:in `method_missing'
```

It looks like you forgot a variable or method here. We also recommend using plain sql to migrate data if it is possible 😄 . You can check some examples here:

https://github.com/decidim/decidim/blob/master/decidim-core/db/migrate/20170713131308_migrate_user_roles_to_participatory_process_roles.rb
https://github.com/decidim/decidim/blob/master/decidim-proposals/db/migrate/20170612101809_migrate_proposals_category.rb

leio10 commented:
Hi @beagleknight, I've changed data migration to plain SQL, fixed migration and rollback, and verified that they are working properly.

beagleknight commented:
Hi @leio10 !

I still seeing some errors on your migration:

```
PG::SyntaxError: ERROR:  syntax error at or near "Anglada"
LINE 3:         SET name = '{"ca":"Ca n'Anglada","es":"Ca n'Anglada"...
```

Is it possible? 😮

leio10 commented:
I forgot to escape single quotes. I fixed it, but I'll use this when I have some time:
 https://stackoverflow.com/questions/4483049/how-to-execute-a-raw-update-sql-with-dynamic-binding-in-rails

beagleknight commented:
I redeployed the test application here: https://decidim-localret-pr-8.herokuapp.com/processes/2/f/8/proposals?utf8=%E2%9C%93&filter%5Bsearch_text%5D=&filter%5Borigin%5D=&filter%5Borigin%5D=all&filter%5Bstate%5D=&filter%5Bstate%5D=all&filter%5Bscope_id%5D%5B%5D=&filter%5Bscope_id%5D%5B%5D=7&filter%5Bcategory_id%5D=&order=random

It seems to work correctly 😄 . I summon @oriolgual and @josepjaume to finish their reviews and we are good to go!

Thanks @leio10 !

deivid-rodriguez commented:
I'm seeing some missing translations:

![screenshot](https://user-images.githubusercontent.com/2887858/28729669-88282f0a-73cd-11e7-85ea-d3a398242ee6.jpg)

Also I see some behavior that I'm not sure it's expected. When put focus on the filter I see all globally defined scopes + the global scope. But when I start filtering, a new option appears.

![weird](https://user-images.githubusercontent.com/2887858/28729642-70a7e91a-73cd-11e7-8553-78ed2ad740f4.gif)

leio10 commented:
Hi @deivid-rodriguez, thanks for the feedback!
Regarding missing translations, I've only added the english texts. As @andreslucena commented in https://github.com/decidim/decidim/pull/1500#issuecomment-310601481, the flow is to add the english translations first, then do the merge and finally add the rest of translations, through Crowdin.
Also, there is a strange issue in the scope names that I saw in your screen captures, but it does not happened to me. It seems a problem of the faker gem, but I don't have clear where is it: https://github.com/stympy/faker/search?l=YAML&q=male_first_name&type=&utf8=%E2%9C%93
Last, I think the scope search behavior is the expected, as we commented in https://github.com/decidim/design/pull/154#issuecomment-311088478, but we can think more about it to find a better solution.

deivid-rodriguez commented:
Thanks @leio10, I reported the missing translations because it was happening for every language, but I didn't realize there was no English in the available languages... So it's fine indeed, sorry about that :sweat_smile:. Regarding the faker thing, I think it's the same problem, faker doesn't have translations for those key/language combinations, should be fine.

Finally, about the usability issue, I think it's a bit counter-intuitive, but I think we can keep it for now and try to improve it in further PR's.

Great work! :clap: :clap: :clap:

deivid-rodriguez commented on decidim-admin/app/controllers/decidim/admin/scope_types_controller.rb:
```diff
@@ -0,0 +1,81 @@
+  module Admin
+    # Controller that allows managing all scopes at the admin panel.
+    #
+    class ScopeTypesController < ApplicationController
```
Can you catch up with #1633 here? Basically fully qualify `ApplicationController` and remove `require_dependency`.

leio10 replied:
Done!

deivid-rodriguez commented on decidim-admin/app/forms/decidim/admin/participatory_process_form.rb:
```diff
@@ -25,7 +25,7 @@ class ParticipatoryProcessForm < Form
       attribute :hashtag, String
       attribute :promoted, Boolean
-      attribute :scope_id, Integer
+      attribute :scope_id, String
```
Is this change necessary?

deivid-rodriguez replied:
I'm seeing that sometimes you will receive the string "global" here, and thus you rely on `global.to_i` being `0` and the ActiveRecord scope being empty (which is what we want, I guess). Is that it?

leio10 replied:
Yes, at first I was using the global scope as an option, so I did what you said. But, finally I changed it to use global scope as the empty selection, which I think is more elegant. But I forgot to rollback the code change that you are commenting. I've reviewed it and I think now it is consistent.

leio10 commented:
@deivid-rodriguez @beagleknight Thank you very much for your review and comments!!!
It's hard to see some silly mistakes I've made, but I hope it will help me to triple check my changes before committing in the future :see_no_evil:

deivid-rodriguez commented on decidim-core/app/helpers/decidim/organization_scopes_helper.rb:
```diff
@@ -27,7 +27,7 @@ def organization_scopes(organization = current_organization)
     # those elements that do not have a scope, you cannot send an empty String or
     # nil because Searchlight will ignore the parameter. The only solution here is
-    # to use a specific String as ID (`"global"`, in our case), and then treat this
+    # to use a specific Numbre as ID (`0`, in our case), and then treat this
```
Typo. This related to what we are discussing in https://github.com/decidim/decidim/pull/1500#discussion_r130640634, isn't it? Is it expected that this changes in `OrganizationScopesHelper`, but the string "global" (or the empty string) is still used in `ParticipatoryProcessScopesHelper`?

leio10 replied:
You are right, I should have deleted that helper, which is no longer used.

deivid-rodriguez commented:
@leio10 No problem! Sorry if I nitpick too much. I tend to nitpick too much. :sweat_smile:

oriolgual commented:
@leio10 could you please rebase this against master?

leio10 commented:
@oriolgual done, but it seems that there is some conflict with sass gem version :confused:

leio10 commented:
@oriolgual Done!

deivid-rodriguez commented:
Looks good! Codeclimate is having some trouble today: http://status.codeclimate.com/, but I anticipate it's going to fail since there's some trailing whitespace in this PR.

You can look for it and fix it using `git grep -I --name-only -e ' $'`, or even better installing the [codeclimate CLI](https://github.com/codeclimate/codeclimate) and running `codeclimate analyze -e grep`.

leio10 commented:
@deivid-rodriguez Yes, I forgot to run `rubocop` before committing. I'm also getting this message:

```
decidim-dev/lib/decidim/dev/test/rspec_support/translation_helpers.rb:21:3: W: Unnecessary disabling of Style/PredicateName.
  # rubocop:disable Style/PredicateName
  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```

I can remove that, but I don't touched that file. Should I fix it?

deivid-rodriguez commented:
Mmm, weird that codeclimate missed that... ¯\_(⊙︿⊙)_/¯. Thanks for fixing it!

The whitespace check is done globally (not only for ruby files), so you need to use the commands I pointed out (or manually look at the diff... :))

deivid-rodriguez commented on decidim-budgets/app/views/decidim/budgets/projects/_filters.html.erb:
```diff
@@ -12,10 +12,10 @@
+    <%= form.scopes_select :scope_id, legend_title: t(".scopes"), label: false, prompt: t('.scope_prompt'), remote_path: decidim.scopes_search_path(include_root: true) %>
   <% end %>
-
+  
```
This would be one of the whitespace offenses being introduced.

deivid-rodriguez replied:
Not sure if there are more.

leio10 commented:
@deivid-rodriguez Ok, I fixed those too. I think codeclimate is not catching those offenses (at least, it didn't catch them before). 

Also, with the `git grep` command I'm getting offenses on files that I didn't touched. Most of them are vendor files, so I only fixed `.babelrc`. This is the full list:
```
.babelrc
decidim-api/vendor/assets/javascripts/decidim/api/graphql-docs.js
decidim-api/vendor/assets/javascripts/decidim/api/react.js
decidim-core/vendor/assets/javascripts/foundation-datepicker.js
decidim-core/vendor/assets/javascripts/leaflet-svg-icon.js
```

deivid-rodriguez commented:
Yes, it's because I only introduced this check recently, so it didn't bite until you rebased...

The exclusions are fine indeed (vendor files are excluded from the check). `codeclimate analyze -e grep` takes the exclusions into account, so that command should be green now.

Thanks for fixing these issues, it might actually prevent some conflicts when I have to rebase my PR's on top of this one after it's merged! :smiley:

josepjaume commented:
This is finally going in!!! Thanks for the super-hard work Y'all! @leio10 @beagleknight @deivid-rodriguez 🕺 💃

beagleknight commented:
Hi @leio10 ! I am sorry but I totally forgot until know 😅 . Can you create a new PR updating the `CHANGELOG.md` file? You should add something to the `Added` section. Thanks!
